File: agent/runtime/active.py
from __future__ import annotations

import asyncio
import time
from collections.abc import Awaitable
from dataclasses import dataclass
from typing import Any, Literal


ActiveTaskKind = Literal["turn", "scheduler", "team", "watchlist"]


@dataclass
class ActiveTaskInfo:
    id: str
    kind: ActiveTaskKind
    label: str
    started_at: float
    turn_id: str | None = None
    job_id: str | None = None
    cancelled: bool = False

    def to_dict(self) -> dict[str, Any]:
        return {
            "id": self.id,
            "kind": self.kind,
            "label": self.label,
            "startedAt": self.started_at,
            "turnId": self.turn_id,
            "jobId": self.job_id,
            "cancelled": self.cancelled,
        }


@dataclass
class _ActiveTask:
    info: ActiveTaskInfo
    task: asyncio.Task


class ActiveTaskRegistry:
    """Process-local registry for user-visible work that can be cancelled."""

    def __init__(self) -> None:
        self._tasks: dict[str, _ActiveTask] = {}
        self._lock = asyncio.Lock()

    async def run(
        self,
        *,
        task_id: str,
        kind: ActiveTaskKind,
        label: str,
        awaitable: Awaitable[Any],
        turn_id: str | None = None,
        job_id: str | None = None,
    ) -> Any:
        info = ActiveTaskInfo(
            id=task_id,
            kind=kind,
            label=label,
            started_at=time.time(),
            turn_id=turn_id,
            job_id=job_id,
        )
        async with self._lock:
            if task_id in self._tasks:
                if hasattr(awaitable, "close"):
                    awaitable.close()  # type: ignore[attr-defined]
                raise RuntimeError(f"active task already exists: {task_id}")
            task = asyncio.create_task(awaitable)
            self._tasks[task_id] = _ActiveTask(info=info, task=task)
        try:
            return await task
        finally:
            async with self._lock:
                current = self._tasks.get(task_id)
                if current and current.task is task:
                    self._tasks.pop(task_id, None)

    async def update(self, task_id: str, **fields: Any) -> ActiveTaskInfo | None:
        async with self._lock:
            active = self._tasks.get(task_id)
            if not active:
                return None
            for key, value in fields.items():
                if hasattr(active.info, key):
                    setattr(active.info, key, value)
            return active.info

    async def cancel(
        self,
        *,
        task_id: str | None = None,
        kind: ActiveTaskKind | None = None,
    ) -> list[ActiveTaskInfo]:
        async with self._lock:
            selected = [
                active
                for active in self._tasks.values()
                if (not task_id or active.info.id == task_id)
                and (not kind or active.info.kind == kind)
            ]
            for active in selected:
                active.info.cancelled = True
                active.task.cancel()
            return [active.info for active in selected]

    async def list(self) -> list[ActiveTaskInfo]:
        async with self._lock:
            return [active.info for active in self._tasks.values()]

    async def has_active(self) -> bool:
        async with self._lock:
            return bool(self._tasks)

File: packages/core/src/runtime/active.ts
export type ActiveTaskKind =
  'turn' | 'scheduler' | 'team' | 'watchlist' | 'goal'

export interface ActiveTaskInfo {
  id: string
  kind: ActiveTaskKind
  label: string
  started_at: number
  turn_id: string | null
  job_id: string | null
  session_id: string | null
  cancelled: boolean
}

export class CancelledTaskError extends Error {
  constructor(taskId: string) {
    super(`active task cancelled: ${taskId}`)
    this.name = 'CancelledTaskError'
  }
}

export class TurnBusyError extends Error {
  readonly code = 'turn_busy'

  constructor() {
    super('Another agent turn is already running')
    this.name = 'TurnBusyError'
  }
}

interface ActiveTask {
  info: ActiveTaskInfo
  cancel: () => void
}

export class ActiveTaskRegistry {
  private readonly tasks = new Map<string, ActiveTask>()

  async run<T>(opts: {
    taskId: string
    kind: ActiveTaskKind
    label: string
    execute: () => Promise<T>
    turnId?: string | null
    jobId?: string | null
    sessionId?: string | null
    abort?: (() => void) | null
  }): Promise<T> {
    if (this.tasks.has(opts.taskId))
      throw new Error(`active task already exists: ${opts.taskId}`)
    let rejectCancel: (error: Error) => void = () => {}
    const cancelPromise = new Promise<never>((_, reject) => {
      rejectCancel = reject
    })
    const info: ActiveTaskInfo = {
      id: opts.taskId,
      kind: opts.kind,
      label: opts.label,
      started_at: Date.now() / 1000,
      turn_id: opts.turnId ?? null,
      job_id: opts.jobId ?? null,
      session_id: opts.sessionId ?? null,
      cancelled: false,
    }
    this.tasks.set(opts.taskId, {
      info,
      cancel: () => {
        info.cancelled = true
        opts.abort?.()
        rejectCancel(new CancelledTaskError(opts.taskId))
      },
    })
    try {
      const awaitable = opts.execute()
      return await Promise.race([awaitable, cancelPromise])
    } finally {
      const current = this.tasks.get(opts.taskId)
      if (current?.info === info) this.tasks.delete(opts.taskId)
    }
  }

  update(
    taskId: string,
    fields: {
      turnId?: string | null
      jobId?: string | null
      sessionId?: string | null
      label?: string | null
    },
  ): ActiveTaskInfo | null {
    const active = this.tasks.get(taskId)
    if (!active) return null
    if (fields.turnId !== undefined) active.info.turn_id = fields.turnId
    if (fields.jobId !== undefined) active.info.job_id = fields.jobId
    if (fields.sessionId !== undefined)
      active.info.session_id = fields.sessionId
    if (fields.label !== undefined && fields.label !== null)
      active.info.label = fields.label
    return active.info
  }

  cancel(
    opts: { taskId?: string | null; kind?: ActiveTaskKind | null } = {},
  ): ActiveTaskInfo[] {
    const selected = [...this.tasks.values()].filter((active) => {
      return (
        (!opts.taskId || active.info.id === opts.taskId) &&
        (!opts.kind || active.info.kind === opts.kind)
      )
    })
    for (const active of selected) active.cancel()
    return selected.map((active) => active.info)
  }

  list(): ActiveTaskInfo[] {
    return [...this.tasks.values()].map((active) => active.info)
  }

  hasActive(): boolean {
    return this.tasks.size > 0
  }

  hasActiveKind(kind: ActiveTaskKind): boolean {
    return [...this.tasks.values()].some((active) => active.info.kind === kind)
  }
}

export function activeTaskToDict(
  info: ActiveTaskInfo,
): Record<string, unknown> {
  return {
    id: info.id,
    kind: info.kind,
    label: info.label,
    startedAt: info.started_at,
    turnId: info.turn_id,
    jobId: info.job_id,
    sessionId: info.session_id,
    session_id: info.session_id,
    cancelled: info.cancelled,
  }
}
